jobs/geo/core.py

```diff
@@ -1,44 +1,90 @@
 from __future__ import annotations
 
-from io import StringIO
-
 from jobs.utils.snowflake import gdf_to_snowflake, snowflake_connection_from_environment
 
 
-def load_geo_data(conn, url: str, table: str) -> None:
-    """Load general geospatial data. Given a URL, load geospatial data into Snowflake."""
-    import geopandas
-    import requests
+def load_ca_geo_data(conn, year: str) -> None:
+    """Load CA Census geo data into Snowflake."""
+    from pygris import (
+        block_groups,
+        blocks,
+        counties,
+        county_subdivisions,
+        places,
+        primary_secondary_roads,
+        pumas,
+        tracts,
+    )
+
+    print(f"Downloading data for CA in year {year}")
+
+    state = "CA"
+    cache = True
+    dfs = []
+
+    county = counties(state=state, year=year, cache=cache)
+    county["year"] = year
+    dfs.append(county)
+
+    tract = tracts(state=state, year=year, cache=cache)
+    tract["year"] = year
+    dfs.append(tract)
 
-    print(f"Downloading data {table} from {url}")
+    block_group = block_groups(state=state, year=year, cache=cache)
+    block_group["year"] = year
+    dfs.append(block_group)
 
-    # Some of the target sources come from Esri servers, and GDAL's vsi system
-    # can be cranky about reading from them (issues with range requests and other
-    # weird errors). So instead we load the data first into a text blob using
-    # requests, then hand it off to geopandas.
-    f = StringIO(requests.get(url).text)
-    gdf = geopandas.read_file(f)
+    block = blocks(state=state, year=year, cache=cache)
+    block["year"] = year
+    dfs.append(block)
 
-    print(f"Writing {table} to snowflake")
-    gdf_to_snowflake(
-        gdf,
-        conn,
-        table_name=table,
-        cluster=False,
+    place = places(state=state, year=year, cache=cache)
+    place["year"] = year
+    dfs.append(place)
+
+    puma = pumas(state=state, year=int(year), cache=cache)
+    puma["year"] = year
+    dfs.append(puma)
+
+    county_subdivision = county_subdivisions(state=state, year=year, cache=cache)
+    county_subdivision["year"] = year
+    dfs.append(county_subdivision)
+
+    primary_secondary_road = primary_secondary_roads(
+        state=state, year=year, cache=cache
     )
+    primary_secondary_road["year"] = year
+    dfs.append(primary_secondary_road)
+
+    table_names = [
+        "COUNTIES_" + year,
+        "TRACTS_" + year,
+        "BLOCK_GROUPS_" + year,
+        "BLOCKS_" + year,
+        "PLACES_" + year,
+        "PUMAS_" + year,
+        "COUNTY_SUBDIVISIONS_" + year,
+        "PRIMARY_SECONDARY_ROADS_" + year,
+    ]
```

Comment:
This is quite fragile: if you add anything or reorder the imports then tables will be written under the wrong name in the destination.

I'd recommend instead storing `dfs` in a dictionary, mapping table name to dataframe. You can then iterate over `dfs.items()` to get the name and the frame, and it will no longer be dependent on the order that you happened to write them in.

Reply:
In fact, we can probably reduce the duplication of the dataframe creation above by creating a mapping from table name to pygris function, then iterating over that to create a mapping from table name to dataframe. Or even doing it in one pass. Pseudocode

```python
loaders = {
  f"TRACTS_{YEAR}": tracts,
  f"BLOCKS_{YEAR}": blocks,
  ... # etc
}
for table_name, loader in loaders.items():
  df = loader(state=state, year=year)
  df["year"] = year
  gdf_to_snowflake(df, conn, ...)
```

Reply:
Thank you for the sample! I knew I could do this with a dictionary, but couldn't remember exactly how. Decided to go with what I knew in favor of getting data loaded faster, but glad for the opportunity to refactor and re-learn this method.

Reply:
done